src/ipnet.rs

```diff
@@ -0,0 +1,828 @@
+    {
+        serializer.serialize_str(&format!("{}", self))
+    }
+}
+
+#[cfg(feature = "schemars")]
+const IPV4_NET_REGEX: &str = concat!(
+    r#"^(([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}"#,
+    r#"([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])"#,
+    r#"/([0-9]|1[0-9]|2[0-9]|3[0-2])$"#,
+);
+
+#[cfg(feature = "schemars")]
+impl schemars::JsonSchema for Ipv4Net {
+    fn schema_name() -> String {
+        "Ipv4Net".to_string()
+    }
+
+    fn json_schema(_: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+        schemars::schema::SchemaObject {
+            metadata: Some(Box::new(schemars::schema::Metadata {
+                title: Some("An IPv4 subnet".to_string()),
+                description: Some("An IPv4 subnet, including prefix and prefix length".to_string()),
+                examples: vec!["192.168.1.0/24".into()],
+                ..Default::default()
+            })),
+            instance_type: Some(schemars::schema::InstanceType::String.into()),
+            string: Some(Box::new(schemars::schema::StringValidation {
+                pattern: Some(IPV4_NET_REGEX.to_string()),
+                ..Default::default()
+            })),
+            extensions: crate::schema_util::extension("Ipv4Net", "0.1.0"),
+            ..Default::default()
+        }
+        .into()
+    }
+}
+
+/// The highest value for an IPv6 subnet prefix
+pub const IPV6_NET_WIDTH_MAX: u8 = 128;
+
+/// An IPv6 subnet
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Ipv6Net {
+    addr: Ipv6Addr,
+    width: u8,
+}
+
+impl Ipv6Net {
+    /// Create an Ipv6Net with the given base address and prefix width.
+    pub fn new(addr: Ipv6Addr, width: u8) -> Result<Self, IpNetPrefixError> {
+        if width > IPV6_NET_WIDTH_MAX {
+            Err(IpNetPrefixError(width))
+        } else {
+            Ok(Self { addr, width })
+        }
+    }
+
+    /// Create an Ipv6Net that contains *exclusively* the given address.
+    pub fn host_net(addr: Ipv6Addr) -> Self {
+        Self {
+            addr,
+            width: IPV6_NET_WIDTH_MAX,
+        }
+    }
+
+    /// Return the base address used to create this subnet.
+    pub fn addr(&self) -> Ipv6Addr {
+        self.addr
+    }
+
+    /// Return the prefix width.
+    pub fn width(&self) -> u8 {
+        self.width
+    }
+
+    pub(crate) fn mask(&self) -> u128 {
+        u128::MAX << (IPV6_NET_WIDTH_MAX - self.width)
+    }
+
+    /// Return the netmask address derived from prefix length.
+    pub fn mask_addr(&self) -> Ipv6Addr {
+        Ipv6Addr::from(self.mask())
+    }
+
+    /// Return true iff the subnet contains only the base address i.e. the
+    /// size is exactly one address.
+    pub fn is_host_net(&self) -> bool {
+        self.width == IPV6_NET_WIDTH_MAX
+    }
+
+    /// Return the number of addresses contained within this subnet.
+    pub fn size(&self) -> u128 {
+        1u128 << (IPV6_NET_WIDTH_MAX - self.width)
```

Comment:
Same comment about overflow, and probably delegating this to bitwise complement of `mask()` plus 1.

Reply:
Interesting. I'm not sure this interface is valid. If we have a /0 then I think the "size" is u128::MAX + 1

What do you think we should do here?

Reply:
I think I'll return an Option